components/ui/project-form.tsx

```diff
@@ -11,56 +11,80 @@ import { useZodForm } from '@/hooks/use-zod-form';
 
 import { addProject, editProject } from '@/app/projects/actions';
 
+import { type Project } from '@prisma/client';
+
 interface Props {
   defaultValues?: Project;
 }
 
-export const schema = z.object({
+const schema = z.object({
   id: z.string().optional(),
   name: z.string().min(1, { message: 'Please enter project name' }),
   upworkId: z.string().nullable(),
   hubstaffId: z.string().nullable(),
   asanaId: z.string().nullable(),
 });
 
-export default function ProjectForm({ defaultValues }: Props) {
+export default function AddProjectForm({ defaultValues }: Props) {
+  const [isPending, startTransition] = useTransition();
+
+  const { execute: executeAdd, status: addStatus } = useAction(addProject, {
+    onSuccess: (data) => {
+      if (data && 'failure' in data) {
+        console.log(data.failure);
+        return;
+      }
+
+      console.log('Project was successfully added'); // fix: won't be printed
+    },
+  });
+
+  const { execute: executeEdit, status: editStatus } = useAction(editProject, {
+    onSuccess: (data) => {
+      if (data && 'failure' in data) {
+        console.log(data.failure);
+        return;
+      }
+
+      console.log('Project was successfully edited'); // fix: won't be printed
+    },
+  });
+
   const {
     register,
     handleSubmit,
     formState: { errors, isValid, isSubmitting, isDirty },
   } = useZodForm({
     schema,
+    mode: 'onBlur',
     defaultValues: defaultValues || {
       name: '',
       upworkId: '',
       hubstaffId: '',
       asanaId: '',
     },
-    mode: 'onBlur',
-  });
-  const router = useRouter();
-
-  const onSubmit = handleSubmit((data) => {
-    if (defaultValues) {
-      const modifiedData = {
-        ...data,
-        id: defaultValues?.id,
-      };
-      editProject(modifiedData)
-        .catch(() => console.error('project edit error'))
-        .then(() => console.log('successfully edited'))
-        .catch(() => console.log('promise catched'));
-    } else {
-      addProject(data)
-        .catch(() => console.error('project add error'))
-        .then(() => console.log('successfully added'))
-        .catch(() => console.log('promise catched'));
-    }
-    router.push('/projects');
   });
 
   return (
-    <form onSubmit={onSubmit} className="paper flex flex-col">
+    <form
+      className="flex flex-col"
+      onSubmit={handleSubmit((data) => {
+        if (defaultValues) {
+          const modifiedData = {
+            ...data,
+            id: defaultValues?.id,
+          };
+
+          startTransition(() => {
+            executeEdit(modifiedData);
+          });
```

Comment:
Since our action already does `redirect`, the `redirect` waiting time is included into `execute`. If you are using `startTransition` only for `isPending` status, it's not needed here. (it would've been useful if we did `router.push()` on the client). Pretty sure action itself is non-blocking, so transition here can be removed